elyra/pipeline/kfp/kfp_authentication.py

```diff
@@ -0,0 +1,693 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC
+from abc import abstractmethod
+from enum import Enum
+from http import HTTPStatus
+import os
+import re
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from urllib.parse import urlsplit
+
+from kfp.auth import KF_PIPELINES_SA_TOKEN_ENV
+from kfp.auth import KF_PIPELINES_SA_TOKEN_PATH
+import requests
+
+
+def _isNoneOrWhitespacesOnly(a_string: str) -> bool:
+    """
+    Utility function: evaluates whether a_string is None or contains
+    only whitespaces.
+
+    :param a_string: string to be evaluated
+    :type string: str
+    :return: True if a_string is None or contains only whitespaces
+    :rtype: Boolean
+    """
+    if a_string is None or len(a_string.strip()) == 0:
+        return True
+    return False
+
+
+class SupportedAuthProviders(Enum):
+    """
+    List of supported authentication providers that is defined
+    in this module. Each entry in this list must be associated
+    with an implementation of AbstractAuthenticator. Each
+    entry's value must also be specified as enum value for the
+    "auth_type" property in elyra/kfp/metadata/schemas/kfp.json
+    to enable users to select the provider from the list.
+
+    """
+    # KF is not secured
+    # (See NoAuthenticationAuthenticator)
+    NO_AUTHENTICATION = 'No authentication'
+    # KF is secured using DEX with static id/password
+    # (See StaticPasswordKFPAuthenticator implementation)
+    DEX_STATIC_PASSWORDS = 'DEX with static passwords'
+    # KF is secured using K8S_SERVICEACCOUNT_TOKEN
+    # (See K8sServiceAccountTokenAuthenticator implementation)
+    K8S_SERVICEACCOUNT_TOKEN = 'Kubernetes service account token'
+    # Supports multiple authentication mechanisms
+    # (See DEXLegacyAuthenticator implementation)
+    DEX_LEGACY = 'DEX (generic)'
+    # Supports DEX with LDAP authentication
+    # (See DEXLDAPAuthenticator implementation)
+    DEX_LDAP_CONNECTOR = 'DEX with LDAP'
```

Comment:
nit: I think a more natural ordering for these is:

- `NO_AUTHENTICATION`
- `K8S_SERVICEACCOUNT_TOKEN`
- `DEX_LDAP_CONNECTOR`
- `DEX_STATIC_PASSWORDS`
- `DEX_LEGACY`